nixos/tests: add integration tests for terminal emulators

JJJollyjim opened the pull request:
###### Motivation for this change

The terminal emulator is a pretty key program for a lot of (most?) desktop users, so it seems nice to have some certainty that they work before updating your system.

Terminal emulators all present a pretty similar interface, which gives us an opportunity to easily test their basic functionality with a single codebase.

My goal here was consistency -- I believe these tests should strike a good balance of thoroughness and simplicity, by not doing text input to the terminals, thereby avoiding much possibility of race conditions or focus issues.

There are two tests run on each terminal emulator
 - can it successfully execute a command passed on the cmdline?
 - can it successfully display a colour?

The latter is used as a proxy for "can it display text?", without going through all the intricacies of OCR.

256-colour terminal mode is used to display the test colour, since it has a universally-applicable palette (unlike 8- and 16- colour, where the colours are implementation-defined), and it is widely supported (unlike 24-bit colour).

I definitely don't have coverage of every terminal emulator -- I will address these in future PRs

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))

~~(note: termite fails to build pending #103704)~~ fixed :+1:

JJJollyjim committed:
nixos/tests: add integration tests for terminal emulators

JJJollyjim committed:
nixos/tests: add passthru.tests to all tested terminal emulators

JJJollyjim committed:
vte: add tested vte-based terminals to passthru.tests

SuperSandro2000 commented:
Can you solve the merge conflicts?

ttuegel commented:
This is a cool idea! I don't feel comfortable reviewing the entire pull request, but for my part: the changes to Konsole look good! :+1:

ttuegel commented:
This is still very useful, just a little stalled.

JJJollyjim commented:
Done! Merge conflicts resolved, references to renamed packages fixed, python import errors fixed, all tests pass.

JJJollyjim commented:
Ready for merging if you're willing @ttuegel @SuperSandro2000

SuperSandro2000 commented on pkgs/applications/terminal-emulators/xterm/default.nix:
```diff
@@ -76,7 +76,7 @@ stdenv.mkDerivation rec {
 
   passthru = {
-    tests = { inherit (nixosTests) xterm; };
+    tests = { customTest = nixosTests.xterm; standardTest = nixosTests.terminal-emulators.xterm; };
```
```suggestion
    tests = {
      customTest = nixosTests.xterm;
      standardTest = nixosTests.terminal-emulators.xterm;
    };
```

JJJollyjim commented:
@SuperSandro2000 I've rebased this and added several remaining missing emulators.

JJJollyjim committed:
nixos/tests/terminal-emulators: fix test for st

some change in the last 24 hours altered the behaviour of st such that
it now dies with a non-zero exit code when the shell exits, so kill is
now necessary

JJJollyjim commented:
/marvin opt-in
/status needs_reviewer

SuperSandro2000 commented on pkgs/applications/terminal-emulators/rxvt-unicode/wrapper.nix:
```diff
@@ -52,6 +53,8 @@ let
         '';
 
         passthru.plugins = plugins;
+
+        passthru.tests.test = nixosTests.terminal-emulators.urxvt;
```
```suggestion
        passthru = {
          plugins = plugins;
          tests.test = nixosTests.terminal-emulators.urxvt;
        };
```

SuperSandro2000 commented on pkgs/applications/terminal-emulators/kitty/default.nix:
```diff
@@ -91,6 +92,8 @@ buildPythonApplication rec {
     })
   ];
 
+  passthru.tests.test = nixosTests.terminal-emulators.kitty;
+
```
Please move this after line 180 where it would be commonly expected.

SuperSandro2000 commented:
Please fix those two little things and then I am just going to merge this and hope for the best.

JJJollyjim commented:
@SuperSandro2000 updated the commit with those changes incorporated, cheers

SuperSandro2000 commented:
Here goes nothing! Thanks.
